rfcs/003-strings.md

```diff
@@ -0,0 +1,325 @@
+
+* `String`: an owned pointer to UTF-8 encoded data together with a known length and capacity.
+* `&'a str`: a view into UTF-8 encoded data that lasts for the lifetime `'a`.
+* `CString`: an owned, C-compatible, nul-terminated string of bytes.
+* `&'a CStr`: a view into a `CString` data that lasts for lifetime `'a`.
+* `OsString` (on Windows): an owned sequence of 8-bit values, encoded in a less-strict variant of UTF-8.
+* `OsStr` (on Windows): a view into `OsString` data that lasts for lifetime `'a`.
+
+#### Equivalences
+
+Windows' `PCSTR` has a direct in-memory equivalent to Rust's `CString` and `&CStr` where the Rust types have clear ownership semantics while the Windows version does not.
+
+### Modeling Windows String types
+
+#### Ownership for Win32 Strings
+
+Often when dealing with Win32 strings (i.e., `PSTR`, `PCSTR`, `PWSTR`, and `PCWSTR`) who owns the data or how long the data is valid for is well known:
+
+* As "IN" params to a function, the string should only be considered valid for the lifetime of the function call.
+* As string literals, the string data is valid for the entirety of the program.
+
+However, sometimes we don't know the ownership or lifetime of a value outside of documentation:
+
+* As "OUT" params to a function, the caller is sometimes owns the data, but it may also be owned by someone else and thus only valid for some lifetime.
+  * [Example of a "borrowed" OUT param](https://microsoft.github.io/windows-docs-rs/doc/windows/Win32/Globalization/struct.IOptionDescription.html#method.Id). The `id` string is owned by the `IOwnDescription` object.
+  * [Example of an "owned" OUT param](https://microsoft.github.io/windows-docs-rs/doc/windows/Win32/System/Com/struct.IEnumString.html#method.Next). The caller is responsible for freeing the `rgelt` param.
+* As fields in a struct, the ownership of the data is unclear. The string can sometimes be freed when the struct is no longer needed but sometimes something else owns that string.
+
+**In short**: there is no way to programmatically know the lifetime of a win32 string except when it is used as an "IN" param.
+
+#### Ownership for `BSTR`
+
+`BSTR` has the same story as a win32 string type.
+
+#### Ownership for `HSTRING`
+
+Ownership for `HSTRING` is fairly clear:
+
+* As "IN" params, the HSTRING is only being borrowed.
+* Typically in any other case, the `HSTRING` handle has had its reference counted incremented and when that handle is no longer needed, the reference count should be decremented.
+
+**In short**: `HSTRING`s are always valid when in scope, and when they are no longer used, their reference counts should be decremented **except** in the case of "IN" params where they are logically "borrowed".
+
+### Proposed API
+
+It's clear that "IN" params are the scenario where ownership differs from the status quo. Therefore, the `windows` crate will treat the types as the following.
+
+* Introduce a new type `CWString` which is the wide string equivalent to `CString`.
```

Comment:
I'll review more carefully when I get a moment but my most obvious question is what `CWString` gives you that `HSTRING` doesn't already provide. Obviously, there are already a dizzying number of string types and I'd love to not make that dizzier. 😉